File: apps/ios/Voiid/Voiid/Games/TicTacToeBoard.swift
//
//  TicTacToeBoard.swift
//  Voiid
//
//  The 3x3 board, shared by the online game (server state) and the bot game (local state).
//
//  Extracted so the two modes cannot drift visually. It takes plain values, not an engine:
//  it has no idea whether the marks came from a server frame or a minimax search, which is
//  exactly the point — one board, two callers.
//
//  Marks are STROKED PATHS, not SF Symbols: a drawn X/O scales cleanly with the cell and
//  can animate its stroke on, which a font glyph cannot.
//
//  THE BOARD OWNS THE END-OF-MATCH BEATS, not its callers (docs/games/TICTACTOE_WIN_LINE.md
//  §2.2). The win line, the swell of the winning cells, the success haptic and the win sound
//  all have to land in a specific order a fifth of a second apart, and a sequence split across
//  two screens is a sequence that drifts. Callers hand over `line` and are told when the
//  gesture has finished, via `onLineComplete`.
//
//  Mirrors Android `TicTacToeBoard.kt`.
//

import SwiftUI

struct TicTacToeBoard: View {
    /// 9 cells, row-major. Each is the seat index that owns it, or nil.
    let board: [Int?]
    /// Winning triple to strike through, if the game is won.
    let line: [Int]?
    /// The match ended with no winner. Drives the draw treatment (§2.3) — deliberately a
    /// separate input rather than "finished && line == nil", because the board cannot see
    /// `finished` and guessing a draw from an absent line would treat every match in progress
    /// as a draw.
    var isDraw: Bool = false
    /// False when taps should be ignored (not your turn, game over, paused).
    let enabled: Bool
    let onTap: (Int) -> Void
    /// Fired when the win stroke has finished drawing — the cue for a caller to reveal its
    /// result panel, so the banner does not land on top of the gesture that explains it.
    var onLineComplete: (() -> Void)?

    /// Reduce-motion draws the line at full length instantly (§2.5). The colour, the sound and
    /// the haptic all survive; only the stroke animation is dropped. The information is the
    /// point — the motion was only ever how it was delivered.
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    /// How much of the win stroke is drawn, 0-1. Animated via `.trim`.
    @State private var strokeProgress: CGFloat = 0
    /// The winning cells swell only once the stroke has landed — it is the payoff now, not
    /// the whole event.
    @State private var swelled = false
    /// Every cell dims together on a draw.
    @State private var drawFaded = false

    // MARK: Timing and geometry (TICTACTOE_WIN_LINE.md §2.1-2.2)
    //
    // KEEP IDENTICAL TO ANDROID. Divergence here is how two ports of one feature end up
    // feeling like different games.

    /// Let the player SEE the third mark before it is annotated. Firing the line on the same
    /// frame as the winning mark makes the two read as one blurred event, and the player never
    /// registers which move won. This is the most important number in the file.
    private static let hold: Double = 0.12
    /// Long enough to read as a deliberate stroke, short enough not to delay the result.
    private static let strokeDuration: Double = 0.34
    /// Extend past the outer two centres by this fraction of a cell, so the line visibly
    /// strikes THROUGH the row rather than connecting two dots. Matters most on the diagonals,
    /// where the un-extended line is visually shortest relative to what it crosses.
    private static let overshoot: CGFloat = 0.35
    /// The line must dominate the marks it crosses, not match them.
    private static let markStrokeWidth: CGFloat = 9
    private static let lineStrokeWidth: CGFloat = markStrokeWidth * 1.4
    private static let drawFadeDuration: Double = 0.30

    var body: some View {
        VStack(spacing: VoiidSpacing.sm) {
            ForEach(0..<3, id: \.self) { row in
                HStack(spacing: VoiidSpacing.sm) {
                    ForEach(0..<3, id: \.self) { col in
                        cell(row * 3 + col)
                    }
                }
            }
        }
        // Overlaid on the GRID, so the geometry reader reports exactly the grid's frame and
        // cell centres fall out of the same spacing constant the layout above uses.
        .overlay { winLine }
        .saturation(drawFaded ? 0.15 : 1)
        .opacity(drawFaded ? 0.65 : 1)
        .onChange(of: line) { _, newLine in runWinSequence(for: newLine) }
        .onChange(of: isDraw) { _, draw in
            withAnimation(reduceMotion ? nil : .easeInOut(duration: Self.drawFadeDuration)) {
                drawFaded = draw
            }
        }
        .onAppear {
            // A board entered ALREADY WON (a rejoin, a returning screen) shows the end state
            // without replaying a gesture the player already missed — and still reports
            // completion, or a caller waiting on `onLineComplete` would wait forever for a
            // sequence that was deliberately skipped.
            if line != nil {
                strokeProgress = 1
                swelled = true
                onLineComplete?()
            }
            drawFaded = isDraw
        }
    }

    // MARK: - Win line

    @ViewBuilder
    private var winLine: some View {
        if let line, line.count == 3 {
            GeometryReader { geo in
                // Cells are square with equal gaps, so one axis defines both — deriving y from
                // the height instead would drift by a rounding error and tilt the diagonals.
                let cell = (geo.size.width - VoiidSpacing.sm * 2) / 3
                let a = Self.centre(of: line[0], cell: cell)
                let b = Self.centre(of: line[2], cell: cell)
                let (start, end) = Self.extended(from: a, to: b, cell: cell)

                WinLineShape(start: start, end: end)
                    .trim(from: 0, to: strokeProgress)
                    .stroke(
                        winnerColour,
                        style: StrokeStyle(lineWidth: Self.lineStrokeWidth, lineCap: .round)
                    )
            }
            .allowsHitTesting(false)
        }
    }

    /// The winner's own mark colour, at full saturation — so the stroke says WHO won in the
    /// same gesture that says that someone won.
    private var winnerColour: Color {
        let seat = line?.first.flatMap { board.indices.contains($0) ? board[$0] : nil }
        return seat == 0 ? VoiidColor.primary : VoiidColor.accent
    }

    private static func centre(of index: Int, cell: CGFloat) -> CGPoint {
        let col = CGFloat(index % 3), row = CGFloat(index / 3)
        return CGPoint(x: col * (cell + VoiidSpacing.sm) + cell / 2,
                       y: row * (cell + VoiidSpacing.sm) + cell / 2)
    }

    private static func extended(from a: CGPoint, to b: CGPoint, cell: CGFloat) -> (CGPoint, CGPoint) {
        let dx = b.x - a.x, dy = b.y - a.y
        let length = max(hypot(dx, dy), 0.0001)
        let ux = dx / length, uy = dy / length
        let reach = cell * overshoot
        return (CGPoint(x: a.x - ux * reach, y: a.y - uy * reach),
                CGPoint(x: b.x + ux * reach, y: b.y + uy * reach))
    }

    /// The end-of-match sequence, in the order §2.2 specifies:
    ///
    ///     0 ms    final mark lands, mark sound plays (the caller's board diff does this)
    ///     120 ms  hold expires — stroke begins, win sound starts with it
    ///     460 ms  stroke complete, winning cells swell, success haptic
    ///     560 ms  caller reveals its result panel
    private func runWinSequence(for newLine: [Int]?) {
        guard newLine != nil else {
            strokeProgress = 0
            swelled = false
            return
        }

        guard !reduceMotion else {
            strokeProgress = 1
            swelled = true
            GameAudio.shared.play("chalk_line", gain: 0.75)
            Haptics.success()
            onLineComplete?()
            return
        }

        DispatchQueue.main.asyncAfter(deadline: .now() + Self.hold) {
            // Sound and stroke start on the SAME beat: the scrape is the stroke, and a scrape
            // that outlasts the line reads as broken without the player knowing why.
            GameAudio.shared.play("chalk_line", gain: 0.75)
            withAnimation(.easeOut(duration: Self.strokeDuration)) { strokeProgress = 1 }
        }
        DispatchQueue.main.asyncAfter(deadline: .now() + Self.hold + Self.strokeDuration) {
            // ON COMPLETION, not on start (§2.4): the haptic is the full stop at the end of
            // the gesture, and firing it at the beginning would confirm a win the player has
            // not been shown yet.
            Haptics.success()
            withAnimation(.spring(response: 0.45, dampingFraction: 0.5)) { swelled = true }
            onLineComplete?()
        }
    }

    // MARK: - Cells

    private func cell(_ index: Int) -> some View {
        let mark = board.indices.contains(index) ? board[index] : nil
        let isWinning = line?.contains(index) ?? false
        let tappable = enabled && mark == nil

        return Button {
            // A REJECTED TAP IS STILL A TAP. Disabling the button meant an occupied cell or a
            // mistimed touch produced nothing at all, which is indistinguishable from the app
            // having missed the finger. The board still refuses the move — it just says so.
            guard tappable else {
                TicTacToeSound.rejected()
                return
            }
            Haptics.tap()
            onTap(index)
        } label: {
            ZStack {
                RoundedRectangle(cornerRadius: VoiidRadius.lg)
                    .fill(isWinning && swelled
                          ? VoiidColor.primary.opacity(0.20)
                          : VoiidColor.surfaceCard)

                if let mark {
                    MarkShape(isCross: mark == 0)
                        .stroke(
                            mark == 0 ? VoiidColor.primary : VoiidColor.accent,
                            style: StrokeStyle(lineWidth: Self.markStrokeWidth,
                                               lineCap: .round, lineJoin: .round)
                        )
                        .padding(VoiidSpacing.lg)
                        // Lands with an overshoot rather than appearing — the bouncy feel.
                        .transition(.scale(scale: 0.4).combined(with: .opacity))
                }
            }
            // SQUARE ON THE STACK, not on the background rectangle.
            //
            // With the ratio on the fill only, the rectangle was square but the ZStack was not —
            // so MarkShape sized itself to the taller stack and drew an X and an O that spilled
            // past the tile and looked stretched. Constraining the stack makes every child square.
            .aspectRatio(1, contentMode: .fit)
            // The winning triple swells — but only AFTER the stroke has been drawn, so it reads
            // as the payoff to the gesture rather than competing with it.
            .scaleEffect(isWinning && swelled ? 1.08 : 1)
        }
        .buttonStyle(.plain)
        // NOT `.disabled(!tappable)` — see the rejected-tap branch above. The guard there is
        // what refuses the move; disabling the button would swallow the touch instead.
        .accessibilityLabel(label(mark: mark, index: index))
        .accessibilityAddTraits(tappable ? [] : .isStaticText)
    }

    private func label(mark: Int?, index: Int) -> String {
        let position = "row \(index / 3 + 1), column \(index % 3 + 1)"
        guard let mark else { return "Empty, \(position)" }
        return "\(mark == 0 ? "X" : "O"), \(position)"
    }
}

/// The sounds a board change makes, derived from the board alone.
///
/// SHARED BY ALL FOUR SCREENS. The two online screens and the two bot screens each used to
/// diff the board themselves and call `mark_x`/`mark_o`; four copies of one rule is four
/// chances for a bot move and a human move to stop sounding identical, which is precisely
/// what a player must not be able to hear.
enum TicTacToeSound {
    /// Play the chalk for whichever mark just landed, plus `catch` if it blocked a threat.
    ///
    /// `mySeat` is the local player's seat, or nil where that is unknown (a bot match's
    /// spectator-less case, where the human is always seat 0).
    static func boardChanged(from old: [Int?], to new: [Int?], mySeat: Int?) {
        guard old.count == new.count else { return }
        guard let i = new.indices.first(where: { old[$0] == nil && new[$0] != nil }),
              let seat = new[i] else { return }

        // A RANDOM VARIANT PER PLACEMENT, not one file with pitch jitter. Chalk is never
        // identical twice and this fires up to nine times in thirty seconds — one file plus
        // jitter still reads as one file by move four.
        GameAudio.shared.playAny(seat == 0 ? GameAudio.chalkX : GameAudio.chalkO, gain: 0.55)

        // §3: "your winning threat gets blocked" is Tic Tac Toe's `catch` moment. Layered
        // UNDER the chalk, never instead of it — the opponent still drew a mark, and the
        // shared sound is the context around it.
        if let mySeat, seat != mySeat, blockedThreat(of: mySeat, at: i, before: old) {
            GameAudio.shared.play(GameAudio.catchShared, gain: 0.5)
        }
    }

    /// Did `cell` complete a line `seat` was one move from taking?
    ///
    /// Pure presentation — it decides which sound to play and nothing else. The server stays
    /// the only referee; this cannot make a match wrong, only under- or over-annotate one.
    private static func blockedThreat(of seat: Int, at cell: Int, before board: [Int?]) -> Bool {
        for line in TicTacToeBot.lines where line.contains(cell) {
            let others = line.filter { $0 != cell }
            if others.allSatisfy({ board[$0] == seat }) { return true }
        }
        return false
    }

    /// An illegal tap: a cell that is taken, or a tap when it is not your turn.
    ///
    /// This had NO feedback at all — the cell was simply not clickable, so a mistimed tap
    /// was indistinguishable from the app not registering the touch. `chalk_stub` is chalk
    /// touching down and never travelling: physically "that didn't take".
    static func rejected() {
        GameAudio.shared.play("chalk_stub", gain: 0.45)
    }
}

/// The stroke through the winning triple. A plain two-point path so `.trim` animates it as a
/// line being DRAWN — a stroke that appears at full length is a graphic, a stroke that is drawn
/// is a gesture, and this is a game about marking a board by hand.
private struct WinLineShape: Shape {
    let start: CGPoint
    let end: CGPoint

    func path(in rect: CGRect) -> Path {
        var p = Path()
        p.move(to: start)
        p.addLine(to: end)
        return p
    }
}

/// An X (two crossing strokes) or an O (a circle), drawn to fill its rect.
private struct MarkShape: Shape {
    let isCross: Bool

    func path(in rect: CGRect) -> Path {
        var p = Path()
        if isCross {
            p.move(to: CGPoint(x: rect.minX, y: rect.minY))
            p.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
            p.move(to: CGPoint(x: rect.maxX, y: rect.minY))
            p.addLine(to: CGPoint(x: rect.minX, y: rect.maxY))
        } else {
            p.addEllipse(in: rect)
        }
        return p
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/games/TicTacToeBoard.kt
package com.voiid.app.main.games

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.EaseOutCubic
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.lerp
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.voiid.app.ui.components.VoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidRadius
import com.voiid.app.ui.theme.VoiidSpacing
import kotlinx.coroutines.delay
import kotlin.math.hypot

/**
 * The 3x3 board, shared by the online game (server state) and the bot game (local state).
 *
 * EXTRACTED so the two modes cannot drift visually — the counterpart to iOS
 * `TicTacToeBoard.swift`, which was extracted for the same reason and left Android's board
 * inline in two screens (docs/games/TICTACTOE.md §2.5). The win line is needed by both screens,
 * which is what finally made two copies untenable.
 *
 * It takes plain values, not an engine: it has no idea whether the marks came from a server
 * frame or a minimax search, which is exactly the point — one board, two callers.
 *
 * THE BOARD OWNS THE END-OF-MATCH BEATS, not its callers (docs/games/TICTACTOE_WIN_LINE.md
 * §2.2). The win line, the swell of the winning cells, the success haptic and the win sound all
 * have to land in a specific order a fifth of a second apart, and a sequence split across two
 * screens is a sequence that drifts. Callers hand over [line] and are told when the gesture has
 * finished, via [onLineComplete].
 *
 * @param board 9 cells, row-major; each is the seat index that owns it, or null.
 * @param line the winning triple to strike through, if the game is won.
 * @param isDraw the match ended with no winner. Deliberately a separate input rather than
 *   "finished && line == null", because the board cannot see `finished` and guessing a draw
 *   from an absent line would treat every match in progress as a draw.
 * @param enabled false when taps should be ignored (not your turn, game over, paused).
 * @param onLineComplete fired when the win stroke has finished drawing — the cue for a caller
 *   to reveal its result panel, so the banner does not land on top of the gesture that
 *   explains it.
 */
@Composable
fun TicTacToeBoard(
    board: List<Int?>,
    line: List<Int>?,
    enabled: Boolean,
    onTap: (Int) -> Unit,
    modifier: Modifier = Modifier,
    isDraw: Boolean = false,
    onLineComplete: (() -> Unit)? = null,
) {
    val context = LocalContext.current
    val haptics = remember(context) { VoiidHaptics(context) }
    // Reduce-motion draws the line at full length instantly (§2.5). The colour, the sound and
    // the haptic all survive; only the stroke animation is dropped. The information is the
    // point — the motion was only ever how it was delivered.
    val reduceMotion = remember(context) { ReduceMotion.isEnabled(context) }

    /** How much of the win stroke is drawn, 0-1. */
    val strokeProgress = remember { Animatable(0f) }
    /** The winning cells swell only once the stroke has landed — the payoff, not the event. */
    var swelled by remember { mutableStateOf(false) }
    // Held in an updated state so the sequence below always calls the CURRENT callback even if
    // the caller recomposed with a new lambda mid-animation.
    val onComplete by rememberUpdatedState(onLineComplete)

    // The end-of-match sequence, in the order §2.2 specifies:
    //
    //     0 ms    final mark lands, mark sound plays (the caller's board diff does this)
    //     120 ms  hold expires — stroke begins, win sound starts with it
    //     460 ms  stroke complete, winning cells swell, success haptic
    //     560 ms  caller reveals its result panel
    LaunchedEffect(line) {
        if (line == null) {
            strokeProgress.snapTo(0f)
            swelled = false
            return@LaunchedEffect
        }
        if (reduceMotion) {
            strokeProgress.snapTo(1f)
            swelled = true
            GameAudio.play("win_line", gain = 0.7f)
            haptics.success()
            onComplete?.invoke()
            return@LaunchedEffect
        }
        // 120 ms hold. Let the player SEE the third mark before it is annotated: firing the
        // line on the same frame as the winning mark makes the two read as one blurred event,
        // and the player never registers which move won. The most important number in the file.
        delay(WIN_LINE_HOLD_MS)
        // Sound and stroke start on the SAME beat: the scrape is the stroke, and a scrape that
        // outlasts the line reads as broken without the player knowing why.
        GameAudio.play("win_line", gain = 0.7f)
        strokeProgress.animateTo(
            1f,
            animationSpec = tween(WIN_LINE_DURATION_MS, easing = EaseOutCubic),
        )
        // ON COMPLETION, not on start (§2.4): the haptic is the full stop at the end of the
        // gesture, and firing it at the beginning would confirm a win the player has not been
        // shown yet.
        haptics.success()
        swelled = true
        onComplete?.invoke()
    }

    // Every cell desaturates together on a draw (§2.3) — a draw is the EXPECTED outcome between
    // competent players and had been sharing the loss treatment, which said nothing.
    val drawFade by animateFloatAsState(
        targetValue = if (isDraw) 1f else 0f,
        animationSpec = tween(if (reduceMotion) 0 else DRAW_FADE_MS),
        label = "drawFade",
    )

    Box(modifier) {
        Column(
            Modifier.fillMaxWidth().alpha(1f - drawFade * 0.30f),
            verticalArrangement = Arrangement.spacedBy(VoiidSpacing.sm),
        ) {
            for (row in 0 until 3) {
                Row(horizontalArrangement = Arrangement.spacedBy(VoiidSpacing.sm)) {
                    for (col in 0 until 3) {
                        val index = row * 3 + col
                        Cell(
                            mark = board.getOrNull(index),
                            isWinning = line?.contains(index) == true && swelled,
                            enabled = enabled && board.getOrNull(index) == null,
                            index = index,
                            fade = drawFade,
                            onTap = { onTap(index) },
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
        }

        // The stroke, drawn OVER the grid and sized to it, so cell centres fall out of the same
        // spacing constant the layout above uses.
        if (line != null && line.size == 3) {
            WinLine(
                line = line,
                board = board,
                progress = strokeProgress.value,
                spacing = VoiidSpacing.sm,
                modifier = Modifier.matchParentSize(),
            )
        }
    }
}

// MARK: Timing and geometry (TICTACTOE_WIN_LINE.md §2.1-2.2)
//
// KEEP IDENTICAL TO iOS `TicTacToeBoard.swift`. Divergence here is how two ports of one feature
// end up feeling like different games.

/** Let the player see the winning mark before it is annotated. */
private const val WIN_LINE_HOLD_MS = 120L
/** Long enough to read as a deliberate stroke, short enough not to delay the result. */
private const val WIN_LINE_DURATION_MS = 340
/**
 * Extend past the outer two centres by this fraction of a cell, so the line visibly strikes
 * THROUGH the row rather than connecting two dots. Matters most on the diagonals, where the
 * un-extended line is visually shortest relative to what it crosses.
 */
private const val WIN_LINE_OVERSHOOT = 0.35f
/** The line must dominate the marks it crosses, not match them. Marks stroke at 0.14 * cell. */
private const val WIN_LINE_WIDTH_SCALE = 1.4f
private const val MARK_STROKE_SCALE = 0.14f
private const val DRAW_FADE_MS = 300
/** Where a drawn match's marks end up: neutral, still legible, no longer anyone's colour. */
private val DRAW_NEUTRAL = Color(0xFF8A8A8A)

/**
 * The stroke through the winning triple.
 *
 * A stroke that appears at full length is a graphic; a stroke that is DRAWN is a gesture, and
 * this is a game about marking a board by hand. Compose has no `.trim`, so the end point is
 * lerped along the line instead — for a straight segment that is exactly equivalent and avoids
 * building a PathMeasure every frame.
 */
@Composable
private fun WinLine(
    line: List<Int>,
    board: List<Int?>,
    progress: Float,
    spacing: Dp,
    modifier: Modifier = Modifier,
) {
    // The winner's own mark colour, at full saturation — so the stroke says WHO won in the same
    // gesture that says that someone won.
    val colour = if (board.getOrNull(line[0]) == 0) VoiidColor.primary else VoiidColor.accent

    Canvas(modifier) {
        val gap = spacing.toPx()
        // Cells are square with equal gaps, so one axis defines both — deriving y from the
        // height instead would drift by a rounding error and tilt the diagonals.
        val cell = (size.width - gap * 2f) / 3f

        fun centre(index: Int) = Offset(
            (index % 3) * (cell + gap) + cell / 2f,
            (index / 3) * (cell + gap) + cell / 2f,
        )

        val a = centre(line[0])
        val b = centre(line[2])
        val dx = b.x - a.x
        val dy = b.y - a.y
        val length = hypot(dx, dy).coerceAtLeast(0.0001f)
        val reach = cell * WIN_LINE_OVERSHOOT
        val start = Offset(a.x - dx / length * reach, a.y - dy / length * reach)
        val end = Offset(b.x + dx / length * reach, b.y + dy / length * reach)

        if (progress <= 0f) return@Canvas
        drawLine(
            color = colour,
            start = start,
            end = Offset(
                start.x + (end.x - start.x) * progress,
                start.y + (end.y - start.y) * progress,
            ),
            strokeWidth = cell * MARK_STROKE_SCALE * WIN_LINE_WIDTH_SCALE,
            cap = StrokeCap.Round,
        )
    }
}

@Composable
private fun Cell(
    mark: Int?,
    isWinning: Boolean,
    enabled: Boolean,
    index: Int,
    /// 0 in play, 1 fully drawn-out. Lerps the mark toward neutral for the draw treatment.
    fade: Float,
    onTap: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // Marks land with an overshoot — the "bouncy" feel. Keyed on the mark so it replays per
    // placement rather than once per screen.
    val pop by animateFloatAsState(
        targetValue = if (mark != null) 1f else 0.4f,
        animationSpec = spring(dampingRatio = 0.42f, stiffness = Spring.StiffnessMediumLow),
        label = "pop",
    )
    // The winning triple swells — but only AFTER the stroke has been drawn, so it reads as the
    // payoff to the gesture rather than competing with it.
    val winScale by animateFloatAsState(
        targetValue = if (isWinning) 1.08f else 1f,
        animationSpec = spring(dampingRatio = 0.35f, stiffness = Spring.StiffnessLow),
        label = "win",
    )

    Box(
        modifier
            .aspectRatio(1f)
            .scale(winScale)
            .clip(RoundedCornerShape(VoiidRadius.lg))
            .background(if (isWinning) VoiidColor.primary.copy(alpha = 0.20f) else VoiidColor.surfaceCard)
            .clickable(enabled = enabled) { onTap() }
            .semantics { contentDescription = ticTacToeCellLabel(mark, index) },
        contentAlignment = Alignment.Center,
    ) {
        if (mark != null) {
            val base = if (mark == 0) VoiidColor.primary else VoiidColor.accent
            val tint = lerp(base, DRAW_NEUTRAL, fade * 0.85f)
            // Drawn rather than icon glyphs: a stroked X and O scale cleanly with the cell and
            // give the crossing-style look that a font glyph cannot. This is also what lets the
            // win line share a visual language with the marks it crosses.
            Canvas(Modifier.fillMaxSize().padding(VoiidSpacing.lg).scale(pop)) {
                val w = size.minDimension
                val width = w * MARK_STROKE_SCALE
                if (mark == 0) {
                    drawLine(tint, Offset(0f, 0f), Offset(w, w), strokeWidth = width, cap = StrokeCap.Round)
                    drawLine(tint, Offset(w, 0f), Offset(0f, w), strokeWidth = width, cap = StrokeCap.Round)
                } else {
                    drawCircle(tint, radius = w / 2f, style = Stroke(width = width, cap = StrokeCap.Round))
                }
            }
        }
    }
}

/** Shared by both Tic Tac Toe screens' accessibility labels. */
internal fun ticTacToeCellLabel(mark: Int?, index: Int): String {
    val position = "row ${index / 3 + 1}, column ${index % 3 + 1}"
    return if (mark == null) "Empty, $position" else "${if (mark == 0) "X" else "O"}, $position"
}
